Cabal/Distribution/Types/VersionRange.hs

```diff
@@ -340,34 +358,94 @@ withinRange v = foldVersionRange
 -- Wildcard range utilities
 --
 
--- | @since 2.2
-wildcardUpperBound :: Version -> Version
-wildcardUpperBound = alterVersion $
-    \lowerBound -> init lowerBound ++ [last lowerBound + 1]
-
 isWildcardRange :: Version -> Version -> Bool
 isWildcardRange ver1 ver2 = check (versionNumbers ver1) (versionNumbers ver2)
   where check (n:[]) (m:[]) | n+1 == m = True
         check (n:ns) (m:ms) | n   == m = check ns ms
         check _      _                 = False
 
--- | Compute next greater major version to be used as upper bound
---
--- Example: @0.4.1@ produces the version @0.5@ which then can be used
--- to construct a range @>= 0.4.1 && < 0.5@
---
--- @since 2.2
-majorUpperBound :: Version -> Version
-majorUpperBound = alterVersion $ \numbers -> case numbers of
-    []        -> [0,1] -- should not happen
-    [m1]      -> [m1,1] -- e.g. version '1'
-    (m1:m2:_) -> [m1,m2+1]
-
 -------------------------------------------------------------------------------
 -- Parsec & Pretty
 -------------------------------------------------------------------------------
 
-instance Pretty VersionRange where
+class Ord v => IsVersion v where
+  dispWild :: v -> Disp.Doc
+  noVersion :: VersionRangeOver v
+  majorUpperBound :: v -> v
```

Comment:
need to preserve haddocks.

Reply:
true; I didn't bother yet, as this was just a quick proof of concept; but haddocks definitely need polish if this PR shall proceed.